src/Nethermind/Nethermind.Consensus/Scheduler/BackgroundTaskScheduler.cs

```diff
@@ -0,0 +1,173 @@
+    private readonly Channel<IActivity> _taskQueue;
+    private readonly ILogger _logger;
+    private readonly IBlockProcessor _blockProcessor;
+    private readonly ManualResetEvent _restartQueueSignal;
+
+    public BackgroundTaskScheduler(IBlockProcessor blockProcessor, int concurrency, ILogManager logManager)
+    {
+        if (concurrency < 1) throw new ArgumentException("concurrency must be at least 1");
+
+        _mainCancellationTokenSource = new CancellationTokenSource();
+        _blockProcessorCancellationTokenSource = new CancellationTokenSource();
+        _taskQueue = Channel.CreateUnbounded<IActivity>();
+        _logger = logManager.GetClassLogger();
+        _blockProcessor = blockProcessor;
+        _restartQueueSignal = new ManualResetEvent(true);
+
+        _blockProcessor.BlocksProcessing += BlockProcessorOnBlocksProcessing;
+        _blockProcessor.BlockProcessed += BlockProcessorOnBlockProcessed;
+
+        for (int i = 0; i < concurrency; i++)
+        {
+            Task.Factory.StartNew(StartChannel, TaskCreationOptions.LongRunning);
+        }
+    }
+
+    private void BlockProcessorOnBlocksProcessing(object? sender, BlocksProcessingEventArgs e)
+    {
+        // On block processing, we cancel the block process cts, causing current task to get cancelled.
+        _blockProcessorCancellationTokenSource.Cancel();
+        // We also reset queue signal, causing it to wait
+        _restartQueueSignal.Reset();
+    }
+
+    private void BlockProcessorOnBlockProcessed(object? sender, BlockProcessedEventArgs e)
+    {
+        // Once block is processed, we replace it with the
+        CancellationTokenSource oldTokenSource = Interlocked.Exchange(ref _blockProcessorCancellationTokenSource, new CancellationTokenSource());
+        oldTokenSource.Dispose();
+        // We also set queue signal causing it to continue queue.
+        _restartQueueSignal.Set();
+    }
+
+
+    private async Task StartChannel()
+    {
+        await foreach (IActivity activity in _taskQueue.Reader.ReadAllAsync(_mainCancellationTokenSource.Token))
+        {
+            if (_blockProcessorCancellationTokenSource.IsCancellationRequested)
+            {
+                // In case of task that is suppose to run when a block is being processed, if there is some time left
+                // from its deadline, we re-queue it. We do this in case there are some task in the queue that already
+                // reached deadline during block processing in which case, it will need to execute in order to handle
+                // its cancellation.
+                if (DateTimeOffset.Now < activity.Deadline)
+                {
+                    await _taskQueue.Writer.WriteAsync(activity, _mainCancellationTokenSource.Token);
+                    // Throttle deque to prevent infinite loop.
+                    await _restartQueueSignal.WaitOneAsync(TimeSpan.FromMilliseconds(1), _mainCancellationTokenSource.Token);
```

Comment:
Could this throw unhandled `OperationCanceledException`? (Is outside the `try`)

Also 1 ms timeout, so should mostly throw if not set?

Reply:
Mostly throw is not set?

Reply:
If set it will just pass through as timeout doesn't apply; if not set with 1ms it will almost always throw striaght away. Just checking that's the behaviour you want (1ms timeout seems strange)

Reply:
Why 1ms will throw? Should it not just pass.

Reply:
Timeout is given to say how long to wait if condition not right; just saying it will only wait for 1ms `TimeSpan.FromMilliseconds(1)` which is almost like not having a timeout at all?

Reply:
And if that's desired could just use an atomic flag if not right and exit the loop rather than throwing?

Reply:
The loop should not exit until the app close. The 1ms is to prevent high CPU usage, it still need to cycle through the channel in case there are tasks that need to execute because it deadlined.

Reply:
But the 1ms isn't a wait time regardless is the time it will wait before setting the task as cancelled, which will then throw a `TaskCancelledException`?

Would be better to use `await Task.Delay(1ms)` if just for the delay; and a simpler structure for the block processing active than WaitHandle?

https://github.com/NethermindEth/nethermind/blob/956c49f78b84c1c7afe870bc72eba9e46f6abad6/src/Nethermind/Nethermind.Core/Extensions/WaitHandleExtensions.cs#L18-L29

Reply:
Oh, it returns a bool 😅

nvm